### PR TITLE
Garbage collection switch

### PR DESCRIPTION
**Pull request by lixinchengdu**
Add a switch to enable/disable garbage collection. This feature could be used  to disable garbage collection for some unit tests.

### CODE REVIEW

File: runtime/src/main/java/org/corfudb/runtime/GarbageInformer.java [hunk 1]
```diff
@@ -22,11 +22,10 @@
 import java.util.Map;
 import java.util.Random;
 import java.util.concurrent.*;
```
Comment (WenbinZhu) [inline runtime/src/main/java/org/corfudb/runtime/GarbageInformer.java]: Expand.

File: runtime/src/main/java/org/corfudb/runtime/GarbageInformer.java [hunk 2]
```diff
@@ -103,6 +124,11 @@ public void stop() {
      * @param locators      A list of locators whose associated SMRRecords are marked as garbage.
      */
     public void addUnsafe(long markerAddress, List<SMRRecordLocator> locators) {
+        // If the function of marking garbage is disable, no garbage decisions are made.
+        if (!rt.getParameters().isGarbageCollectionEnable()) {
```
Comment (WenbinZhu) [inline runtime/src/main/java/org/corfudb/runtime/GarbageInformer.java]: I think you should disable garbage collection in VLO, to not call these APIs at all.

File: runtime/src/main/java/org/corfudb/runtime/GarbageInformer.java (hunk 1 above)
Comment (lixinchengdu) [reply runtime/src/main/java/org/corfudb/runtime/GarbageInformer.java]: done.

File: runtime/src/main/java/org/corfudb/runtime/GarbageInformer.java (hunk 2 above)
Comment (lixinchengdu) [reply runtime/src/main/java/org/corfudb/runtime/GarbageInformer.java]: moved to VLO.

File: runtime/src/main/java/org/corfudb/runtime/GarbageInformer.java [hunk 3]
```diff
@@ -86,14 +106,23 @@ public void start() {
...
+
+        log.info("Garbage informer is started.");
     }
 
     /**
      * Stop to send garbage decisions to LogUnit servers.
      * This function shuts down the single-threaded executor.
      */
     public void stop() {
+        if (!isRunning.get()) {
```
Comment (WenbinZhu) [inline runtime/src/main/java/org/corfudb/runtime/GarbageInformer.java]: I think it's better to use compareAndSet just as what you did in start()

File: test/src/test/java/org/corfudb/runtime/view/AbstractViewTest.java [hunk 4]
```diff
@@ -106,6 +106,7 @@ public AbstractViewTest(boolean followBackpointers) {
         CorfuRuntime.overrideGetRouterFunction = this::getRouterFunction;
         runtime = CorfuRuntime.fromParameters(CorfuRuntimeParameters.builder()
                 .followBackpointersEnabled(followBackpointers)
+                .garbageCollectionEnable(false)
```
Comment (WenbinZhu) [inline test/src/test/java/org/corfudb/runtime/view/AbstractViewTest.java]: You should also check AbstractIT and universe framework to disable gc in their runtime as well.

File: runtime/src/main/java/org/corfudb/runtime/GarbageInformer.java (hunk 3 above)
Comment (lixinchengdu) [reply runtime/src/main/java/org/corfudb/runtime/GarbageInformer.java]: Let's discuss it offline.

**Comment by annym** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 96
- Added 148
           

Complexity increasing per file
==============================
- runtime/src/main/java/org/corfudb/protocols/logprotocol/SMRGarbageRecord.java  3
- infrastructure/src/main/java/org/corfudb/infrastructure/log/StreamLogCompactor.java  13
- infrastructure/src/main/java/org/corfudb/infrastructure/log/StreamLogDataStore.java  2
- runtime/src/main/java/org/corfudb/protocols/logprotocol/SMRLogEntry.java  3
- runtime/src/main/java/org/corfudb/runtime/GarbageInformer.java  6
- runtime/src/test/java/org/corfudb/protocols/logprotocol/SMRGarbageEntryTest.java  1
- runtime/src/main/java/org/corfudb/runtime/view/replication/QuorumReplicationProtocol.java  1
- infrastructure/src/main/java/org/corfudb/infrastructure/log/GarbageSizeFirstPolicy.java  2
- infrastructure/src/main/java/org/corfudb/infrastructure/datastore/DataStore.java  1
- infrastructure/src/main/java/org/corfudb/infrastructure/log/SnapshotLengthFirstPolicy.java  5
- infrastructure/src/main/java/org/corfudb/infrastructure/log/SegmentManager.java  5
- infrastructure/src/main/java/org/corfudb/infrastructure/log/GarbageLogSegment.java  9
- infrastructure/src/main/java/org/corfudb/infrastructure/log/StreamLogSegment.java  8
- infrastructure/src/main/java/org/corfudb/infrastructure/log/ClosedSegmentException.java  1
- test/src/test/java/org/corfudb/infrastructure/log/StreamLogCompactorTest.java  8
- runtime/src/main/java/org/corfudb/protocols/logprotocol/SMRRecord.java  3
- infrastructure/src/main/java/org/corfudb/infrastructure/log/AbstractCompactionPolicy.java  3
- runtime/src/main/java/org/corfudb/protocols/logprotocol/SMRGarbageEntry.java  3
- runtime/src/test/java/org/corfudb/protocols/logprotocol/SMRRecordLocatorTest.java  1
- test/src/test/java/org/corfudb/runtime/object/transactions/SnapshotTransactionContextTest.java  1
- runtime/src/main/java/org/corfudb/runtime/collections/MapLocatorStore.java  2
- infrastructure/src/main/java/org/corfudb/infrastructure/log/CompactionMetadata.java  5
- infrastructure/src/main/java/org/corfudb/infrastructure/log/AbstractLogSegment.java  8
- runtime/src/test/java/org/corfudb/protocols/wireprotocol/LogDataTest.java  1
- runtime/src/main/java/org/corfudb/protocols/logprotocol/SMRRecordLocator.java  4
         

Complexity decreasing per file
==============================
+ test/src/test/java/org/corfudb/infrastructure/log/StreamLogFilesTest.java  -5
+ runtime/src/main/java/org/corfudb/runtime/object/transactions/AbstractTransactionalContext.java  -3
+ infrastructure/src/main/java/org/corfudb/infrastructure/log/LogMetadata.java  -1
+ infrastructure/src/main/java/org/corfudb/infrastructure/log/StreamLogFiles.java  -4
+ runtime/src/main/java/org/corfudb/runtime/view/stream/StreamAddressSpace.java  -3
+ infrastructure/src/main/java/org/corfudb/infrastructure/LogUnitServer.java  -1
+ runtime/src/main/java/org/corfudb/runtime/view/stream/AbstractStreamContext.java  -3
+ infrastructure/src/main/java/org/corfudb/infrastructure/SequencerServer.java  -1
+ test/src/test/java/org/corfudb/integration/WorkflowIT.java  -2
+ runtime/src/main/java/org/corfudb/runtime/view/stream/AbstractQueuedStreamView.java  -5
+ runtime/src/main/java/org/corfudb/runtime/view/stream/AddressMapStreamView.java  -4
+ runtime/src/main/java/org/corfudb/runtime/view/AddressSpaceView.java  -1
+ infrastructure/src/main/java/org/corfudb/infrastructure/log/InMemoryStreamLog.java  -3
+ runtime/src/main/java/org/corfudb/runtime/view/stream/BackpointerStreamView.java  -5
+ runtime/src/main/java/org/corfudb/recovery/FastObjectLoader.java  -1
         

Clones added
============
- test/src/test/java/org/corfudb/infrastructure/log/StreamLogCompactorTest.java  2
- test/src/test/java/org/corfudb/runtime/object/transactions/OptimisticTransactionContextTest.java  1
- test/src/test/java/org/corfudb/runtime/object/transactions/SnapshotTransactionContextTest.java  1
         

Clones removed
==============
+ test/src/test/java/org/corfudb/infrastructure/log/StreamLogFilesTest.java  -2
+ test/src/test/java/org/corfudb/integration/WorkflowIT.java  -2
         
```


See the complete overview on [Codacy](https://app.codacy.com/app/CorfuDB/CorfuDB/pullRequest?prid=4653405&bid=15584312)

**Comment by xnull** [comment]
Obsolete?

**Pull request closed without merging**